### DISCUSSION TITLE
Repeat action w/ selection?

### DISCUSSION THREAD

**Discussion by askreet** [helix-editor/helix#5870] [Ideas]
Hello again! Been using Helix as a daily driver now for a month or so and one vim habit I'm having a hard time finding an efficient replacement for is the semantic difference between `.` in Helix and vim.

In vim many eons ago someone told me to get used to `diw` rather than `dw` even if at the beginning of the word because it makes repeating word deletions faster with `.`. So, I can be in the middle of one word, press `diw` then move my cursor anywhere in another word and press `.` to delete that word too.

In Helix, the `.` repeats the previous command, but doesn't execute any selection action. ~Effectively, it's a shortcut replacing one character with another (unless I've missed something).~ (Edit: After writing this I realized it only repeats inserts and appends, even.)

So I'm wondering:
1. Is there such a thing as "repeat my previous selection action and command"? E.g. if I type `mi"d` to delete the contents of a string, can I have a keypress repeat _that_?
2. Should the `.` command (optionally) do something like that, instead?
3. And, generally, have others moved from vim and found better replacement workflows that make this particular muscle memory I have less useful?

Thanks!

**Comment by kirawi** [comment]
`Alt-.` repeats the last `f`, `t`, or `m` motion

**Reply by askreet** [reply]
Interesting, so assuming my example above `Alt-. .` gets me where I want to be.

Is there a similar repeat key for e.g. `3w` motions? I couldn't find one in `<space>?`.

**Comment by gabydd** [comment]
This is an interesting one because we can use ``Alt-.`` (``repeat_last_motion``) but ``.`` only repeats the last command that brought you into insert and then what you inserted, this means ``d`` doesn't actually get repeated but you could remap ``d`` to ``["change_selection", "normal_mode"]``, you can't have the repeat last motion plus repeat last insert as one command because ``.`` isn't actually a command but a special case in the event handling.

**Comment by lxcode** [comment]
The thing that constantly throws me is that after using ```r```, you can't do the same single-character replacement using ```.```, ```Alt-.``` or any combination. Same with things like ```~```.

Similarly, in vim I'll regularly do something like ```dap``` to delete a paragraph, but in Helix you just do ```vapd``` and then...well, you just can't redo it. This seems like kind of a big deal.

**Comment by pozix604** [comment]
Vim uses imperative sentences, with verb and noun. `dap` is [delete] a [paragraph]. When you repeat it, you repeat the entire imperative.

These are separated in Helix. When you repeat, you just repeat the verb, but the noun (selection) is whatever you happened to have selected last. Breaks the best design feature of Vim.

**Comment by CAESIUS-TIM** [comment]
> This is an interesting one because we can use `Alt-.` (`repeat_last_motion`) but `.` only repeats the last command that brought you into insert and then what you inserted, this means `d` doesn't actually get repeated but you could remap `d` to `["change_selection", "normal_mode"]`, you can't have the repeat last motion plus repeat last insert as one command because `.` isn't actually a command but a special case in the event handling.

Cannot delete the single blank line after binding.

**Comment by rickswe** [comment]
I'm sure the contributors put a lot of thought into this one even though it brings frustrations to many. 
I'm missing the vim '.' feature, not to mention the other features...

**Reply by kirawi** [reply]
https://github.com/helix-editor/helix/issues/501#issuecomment-1869920017

**Reply by rickswe** [reply]
I understand helix is selection first it's just inconvenient since that's what the majority of us 
coming from vim/neovim/evil are use to. Perhaps I would look into a macro or command to
implement this later if ever.

**Comment by soloturn** [comment]
50% of my vi repeats could be done with helix **multiple cursors, with shift-c.** the other 50% i am still searching :)
